# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам
Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода.

Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [125]:
import pandas as pd
# выбор 8 столбцов для изучения
raw = {0:('surgery', 'хирургия'), 1:('age',), 3:('rectal_temperature',), 4:('pulse',),
       6:('temperature_of_extremities', 'температура_конечностей'), 
       10:('pain', 'боль-субъективное суждение об уровне боли лошади'), 
       21:('abdomcentesis_total_protein', 'Общий белок брюшной полости'), 
       22:('outcome', 'что случилось с лошадью')
      }

horse_data = pd.read_csv('D:/DATA_SCIENCE/dataset/14_dz_stat_basics/horse_data.csv', 
                         header=None, na_values='?', usecols=[k for k in raw.keys()], names=[x[0] for x in raw.values()])

horse_data.head(15)

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
0       2.0    1                38.5   66.0                         3.0   5.0   
1       1.0    1                39.2   88.0                         NaN   3.0   
2       2.0    1                38.3   40.0                         1.0   3.0   
3       1.0    9                39.1  164.0                         4.0   2.0   
4       2.0    1                37.3  104.0                         NaN   NaN   
5       2.0    1                 NaN    NaN                         2.0   2.0   
6       1.0    1                37.9   48.0                         1.0   3.0   
7       1.0    1                 NaN   60.0                         3.0   NaN   
8       2.0    1                 NaN   80.0                         3.0   4.0   
9       2.0    9                38.3   90.0                         1.0   5.0   
10      1.0    1                38.1   66.0                         3.0   3.0   
11      2.0    1                39.1   72.0                         2.0   2.0   
12      1.0    1                37.2   42.0                         2.0   3.0   
13      2.0    9                38.0   92.0                         1.0   1.0   
14      1.0    1                38.2   76.0                         3.0   3.0   

    abdomcentesis_total_protein  outcome  
0                           NaN      2.0  
1                           2.0      3.0  
2                           NaN      1.0  
3                           5.3      2.0  
4                           NaN      2.0  
5                           NaN      1.0  
6                           NaN      1.0  
7                           NaN      2.0  
8                           NaN      3.0  
9                           2.2      1.0  
10                          3.6      1.0  
11                          NaN      1.0  
12                          NaN      1.0  
13                          NaN      2.0  
14                          2.0      1.0

In [126]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   age                          300 non-null    int64  
 2   rectal_temperature           240 non-null    float64
 3   pulse                        276 non-null    float64
 4   temperature_of_extremities   244 non-null    float64
 5   pain                         245 non-null    float64
 6   abdomcentesis_total_protein  102 non-null    float64
 7   outcome                      299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [127]:
horse_data.describe()

surgery         age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       temperature_of_extremities       pain  abdomcentesis_total_protein  \
count                  244.000000  245.00000                   102.000000   
mean                     2.348361    2.95102                     3.019608   
std                      1.045054    1.30794                     1.968567   
min                      1.000000    1.00000                     0.100000   
25%                      1.000000    2.00000                     2.000000   
50%                      3.000000    3.00000                     2.250000   
75%                      3.000000    4.00000                     3.900000   
max                      4.000000    5.00000                    10.100000   

          outcome  
count  299.000000  
mean     1.551839  
std      0.737187  
min      1.000000  
25%      1.000000  
50%      1.000000  
75%      2.000000  
max      3.000000

In [128]:
# Pacсчёт моды
horse_data_mode_old = horse_data.round().mode()
horse_data_mode_old

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
0      1.0    1                38.0   48.0                         3.0   3.0   

   abdomcentesis_total_protein  outcome  
0                          2.0      1.0

In [129]:
# Расчёт дисперсии
pd.DataFrame(horse_data.var()).transpose()

surgery       age  rectal_temperature       pulse  \
0  0.240399  4.726154            0.536247  819.708775   

   temperature_of_extremities      pain  abdomcentesis_total_protein   outcome  
0                    1.092137  1.710706                     3.875255  0.543445

Результаты:
   1. Кол-во лошадей (surgery), подвергшихся хирургическому вмешательству больше, чем без него (среднее = 1,39, медиана, мода = 1, 3 кв = 2,00).
   2. Средний возраст лошадей - взрослые лошади = 1. При этом имеются выбросы (ошибки) = 9, что повлияло на расчёт среднего арифметического = 1,64.
   3. Ректальная тем-ра у большинства повышенная: мода  =38, средняя = 38,1.
   4. Средний пульс (pulse) выше нормы (30-40): mean = 71 (может быть связано с max = 184 (ошибка или шок?), наибольшее кол-во лошадей с пульсом 48 (причиной может быть болезнь, шок).
   5. Температура конечностей (temperature_of_extremities):  75% лошадей имеют тем-ру конечностей отличную от нормы (болезненное состояние), наибольшее число лошадей с прохладными конечностями (мода = 3) - болезненное состояние (шок), max тем-ры конечностей коррелирует с max ректальной тем-ры, что соответствует описанию.
   6. Большее число лошадей испытывают умеренную, перемежающую боль pain мода = 3, 25% испытвают сильные боли, возможно требующие хирургического вмешательства.
   7. Общий белок брюшной полости (abdomcentesis_total_protein) прослеживается зависимость увеличения значения, свидетельствующее о нарушении кишечника, с ростом остальных показателей блоезни.
   8. Outcome. mode = 1 показывает, что большинство лошадей жили. 25% лошадей была сделана операция и 25 % лошадей умерло или усыплено 3 квадриль = 2. 
   9. Кол-во результатов (count) различается, что указывает на наличие пропусков (отсутсвие показателей).


Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

1. В столбце age выявлено 24 выброса со значением = 9, которое является ошибочно введённым значением. И нет ни одной записи со значением 2 (молодая лошадь). Принимаем решение о замене 9 на 2. Отмечаем снижение среднеквадратичного отклонения с 2,17 до 0,27.

In [130]:
horse_data.age.replace(9, 2, inplace = True)
horse_data_old = horse_data.describe()
print(pd.DataFrame(horse_data.age.describe()).transpose())

     count  mean       std  min  25%  50%  75%  max
age  300.0  1.08  0.271746  1.0  1.0  1.0  1.0  2.0


2. Найдем остальные выбросы через межквартильный размах

In [131]:
q1 = horse_data.quantile(0.25)
q3 = horse_data.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
# По ректальной тем-ре
remove_outliers = horse_data[horse_data.rectal_temperature.between(lower_bound.rectal_temperature, upper_bound.rectal_temperature, inclusive = True)]
remove_outliers.describe() - horse_data.describe()

surgery        age  rectal_temperature      pulse  \
count -74.000000 -74.000000           -14.00000 -56.000000   
mean    0.010896   0.004071             0.00111  -2.872134   
std     0.002420   0.006363            -0.16034   0.043221   
min     0.000000   0.000000             1.40000   0.000000   
25%     0.000000   0.000000             0.00000   0.000000   
50%     0.000000   0.000000             0.00000  -4.000000   
75%     0.000000   0.000000             0.00000  -2.000000   
max     0.000000   0.000000            -1.30000   0.000000   

       temperature_of_extremities       pain  abdomcentesis_total_protein  \
count                  -59.000000 -58.000000                   -22.000000   
mean                    -0.137550  -0.191662                     0.066642   
std                      0.025243  -0.071897                     0.157209   
min                      0.000000   0.000000                     0.000000   
25%                      0.000000   0.000000                     0.000000   
50%                     -1.000000   0.000000                    -0.100000   
75%                      0.000000   0.000000                     0.200000   
max                      0.000000   0.000000                     0.000000   

         outcome  
count -74.000000  
mean   -0.076284  
std    -0.017206  
min     0.000000  
25%     0.000000  
50%     0.000000  
75%     0.000000  
max     0.000000

In [132]:
remove_outliers.round().mode() - horse_data.round().mode()
# Отмечаем, что значительно изменилась  мода тем-ры конечностей. Удаление выброса может повлиять на оценку. 

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
0      0.0    0                 0.0    0.0                        -2.0   0.0   

   abdomcentesis_total_protein  outcome  
0                          0.0      0.0

In [133]:
#  Выбросы по пульсу
remove_outliers = horse_data[horse_data.pulse.between(lower_bound.pulse, upper_bound.pulse, inclusive = True)]
remove_outliers.describe() - horse_data.describe()

surgery        age  rectal_temperature      pulse  \
count -29.000000 -29.000000          -11.000000  -5.000000   
mean    0.009414  -0.017269           -0.031235  -1.654741   
std     0.001960  -0.028820           -0.004001  -2.548072   
min     0.000000   0.000000            0.000000   0.000000   
25%     0.000000   0.000000            0.000000   0.000000   
50%     0.000000   0.000000           -0.100000   0.000000   
75%     0.000000   0.000000            0.000000   0.000000   
max     0.000000   0.000000            0.000000 -38.000000   

       temperature_of_extremities       pain  abdomcentesis_total_protein  \
count                  -19.000000 -20.000000                    -5.000000   
mean                    -0.028361  -0.075465                     0.014413   
std                      0.000686  -0.021826                    -0.000230   
min                      0.000000   0.000000                     0.900000   
25%                      0.000000   0.000000                     0.000000   
50%                      0.000000   0.000000                    -0.050000   
75%                      0.000000   0.000000                     0.000000   
max                      0.000000   0.000000                     0.000000   

         outcome  
count -29.000000  
mean   -0.003691  
std     0.016673  
min     0.000000  
25%     0.000000  
50%     0.000000  
75%     0.000000  
max     0.000000

In [134]:
remove_outliers.round().mode() - horse_data.round().mode()

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
0      0.0    0                 0.0    0.0                         0.0   0.0   

   abdomcentesis_total_protein  outcome  
0                          0.0      0.0

In [135]:
#  Выбросы по Общему белку брюшной полости
remove_outliers = horse_data[horse_data.abdomcentesis_total_protein.between(lower_bound.abdomcentesis_total_protein, upper_bound.abdomcentesis_total_protein, inclusive = True)]
remove_outliers.describe() - horse_data.describe()

surgery         age  rectal_temperature       pulse  \
count -205.000000 -205.000000         -158.000000 -184.000000   
mean    -0.089483    0.004211            0.005254    6.315217   
std     -0.025949    0.007430            0.122624    0.449516   
min      0.000000    0.000000            0.000000    6.000000   
25%      0.000000    0.000000            0.000000    5.500000   
50%      0.000000    0.000000            0.000000    8.000000   
75%      0.000000    0.000000            0.100000    5.500000   
max      0.000000    0.000000           -0.500000  -20.000000   

       temperature_of_extremities        pain  abdomcentesis_total_protein  \
count                 -167.000000 -161.000000                    -7.000000   
mean                     0.197094    0.275170                    -0.372239   
std                      0.013066   -0.184786                    -0.554892   
min                      0.000000    0.000000                     0.000000   
25%                      0.000000    0.000000                     0.000000   
50%                      0.000000    0.000000                    -0.250000   
75%                      0.000000    0.000000                    -0.300000   
max                      0.000000    0.000000                    -3.500000   

          outcome  
count -205.000000  
mean     0.192841  
std      0.104421  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      0.750000  
max      0.000000

In [136]:
remove_outliers.round().mode() - horse_data.round().mode()

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
0      0.0  0.0                 0.0    4.0                         0.0   0.0   
1      NaN  NaN                 NaN    NaN                         NaN   NaN   
2      NaN  NaN                 NaN    NaN                         NaN   NaN   

   abdomcentesis_total_protein  outcome  
0                          0.0      0.0  
1                          NaN      NaN  
2                          NaN      NaN

In [137]:
pd.concat([horse_data, remove_outliers]).drop_duplicates(keep=False)

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
0        2.0    1                38.5   66.0                         3.0   
2        2.0    1                38.3   40.0                         1.0   
4        2.0    1                37.3  104.0                         NaN   
5        2.0    1                 NaN    NaN                         2.0   
6        1.0    1                37.9   48.0                         1.0   
..       ...  ...                 ...    ...                         ...   
290      2.0    1                38.6   45.0                         2.0   
292      1.0    1                37.0   66.0                         1.0   
294      2.0    1                38.5   40.0                         1.0   
295      1.0    1                 NaN  120.0                         4.0   
297      1.0    1                37.5   72.0                         4.0   

     pain  abdomcentesis_total_protein  outcome  
0     5.0                          NaN      2.0  
2     3.0                          NaN      1.0  
4     NaN                          NaN      2.0  
5     2.0                          NaN      1.0  
6     3.0                          NaN      1.0  
..    ...                          ...      ...  
290   1.0                          NaN      1.0  
292   4.0                          NaN      2.0  
294   2.0                          NaN      1.0  
295   2.0                          NaN      3.0  
297   4.0                          NaN      2.0  

[180 rows x 8 columns]

### Выводы:
    Применяем замену выбросов к столбцу Возраст.
    Остальные выбросы носят естественный характер, связанный вероятно с болезнями лошадей в острой форме с превышением показателей пульса, ректальной тем-ры.
    Удалять данные выбросы нецелесообразно.
    По столбцу Общий белок брюшной полости кол-во пропусков составляет около 2/3, удалять выбросы нецелесообразно.

Задание 3. Работа с пропусками
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [138]:
horse_data.isna().sum()

surgery                          1
age                              0
rectal_temperature              60
pulse                           24
temperature_of_extremities      56
pain                            55
abdomcentesis_total_protein    198
outcome                          1
dtype: int64

In [139]:
def procent_isna(df):
    for col in df.columns:
        pct_missing = df[col].isnull().mean()
        print(f'{col} - {pct_missing :.1%}')
procent_isna(horse_data)

surgery - 0.3%
age - 0.0%
rectal_temperature - 20.0%
pulse - 8.0%
temperature_of_extremities - 18.7%
pain - 18.3%
abdomcentesis_total_protein - 66.0%
outcome - 0.3%


In [140]:
# Проверяем пропуски по Хирургии
horse_data.loc[horse_data.surgery.isnull()].head()

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
132      NaN    1                38.0   48.0                         3.0   

     pain  abdomcentesis_total_protein  outcome  
132   4.0                          1.0      NaN

In [141]:
horse_data.loc[(horse_data.rectal_temperature == 38) & (horse_data.pain == 4.0)].sort_values('age')


surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
132      NaN    1                38.0   48.0                         3.0   
248      1.0    1                38.0  108.0                         2.0   
260      2.0    1                38.0   40.0                         NaN   
276      1.0    1                38.0   36.0                         3.0   

     pain  abdomcentesis_total_protein  outcome  
132   4.0                          1.0      NaN  
248   4.0                          NaN      1.0  
260   4.0                          NaN      1.0  
276   4.0                          1.0      3.0

Обнаружены пропуски в строке 132 столбец Хирургия и Результат.
Решение: Находим группы записей с одинаковыми параметрами rectal_temperature и боль.
В 2-х записях из 3-х с одинаковым уровнем боли было хиругическое вмешательство - заменим NaN на 1.
В 2-х записях из 3-х лошади жили (включая повышенный пульс по строке 248) - заменим NaN на 1.

In [142]:
horse_data.surgery.fillna(1.0, inplace = True)
horse_data.outcome.fillna(1.0, inplace = True)
horse_data.isna().sum()

surgery                          0
age                              0
rectal_temperature              60
pulse                           24
temperature_of_extremities      56
pain                            55
abdomcentesis_total_protein    198
outcome                          0
dtype: int64

Пропуски в столбце abdomcentesis_total_protein занимают 66 %, что делает этот столбец не информативным.
Решение: удалить целиком столбец abdomcentesis_total_protein

In [143]:
horse_data.drop(['abdomcentesis_total_protein'], axis = 1, inplace = True)
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     300 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   temperature_of_extremities  244 non-null    float64
 5   pain                        245 non-null    float64
 6   outcome                     300 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 16.5 KB


По условиям задачи:
с шоком связаны пониженная < 37.8 ректальная тем-ра, повышенный пульс > 40 и пониженная тем-ра конечностей (3, 4);
повышенная тем-ра конечностей  коррелирует с повышенной ректальной температурой
Сгруппируем по этим данным и заменим Ректальную тем-ру (NaN)  и тем-ру конечностей на медиану по группе.

In [144]:
horse_data.rectal_temperature.fillna(horse_data.groupby(['temperature_of_extremities', 'pulse']).rectal_temperature.transform('median'), inplace = True)
horse_data.temperature_of_extremities.fillna(horse_data.groupby(['rectal_temperature']).temperature_of_extremities.transform('median'), inplace = True)


Наличие боли связано с необходимостью хирургического вмешательства
Заменим NaN в боли на медиану сгруппировав по хирургическому вмешательству

In [145]:
horse_data.pain.fillna(horse_data.groupby(['surgery']).pain.transform('median'), inplace = True)
horse_data.loc[horse_data.rectal_temperature.isnull()]

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
5        2.0    1                 NaN    NaN                         2.0   
16       1.0    2                 NaN  128.0                         3.0   
28       1.0    1                 NaN    NaN                         NaN   
35       2.0    1                 NaN  104.0                         4.0   
52       2.0    1                 NaN    NaN                         1.0   
56       1.0    1                 NaN    NaN                         NaN   
58       1.0    1                 NaN    NaN                         4.0   
74       1.0    2                 NaN    NaN                         NaN   
78       1.0    1                 NaN    NaN                         3.0   
87       2.0    1                 NaN   40.0                         NaN   
93       2.0    1                 NaN    NaN                         3.0   
96       2.0    1                 NaN   84.0                         NaN   
105      1.0    1                 NaN   70.0                         1.0   
110      1.0    1                 NaN   70.0                         3.0   
115      2.0    1                 NaN    NaN                         3.0   
136      1.0    1                 NaN    NaN                         3.0   
159      2.0    1                 NaN    NaN                         1.0   
160      1.0    1                 NaN    NaN                         NaN   
169      1.0    1                 NaN   65.0                         NaN   
174      2.0    1                 NaN    NaN                         3.0   
175      1.0    1                 NaN    NaN                         NaN   
177      1.0    1                 NaN   68.0                         NaN   
191      2.0    2                 NaN  100.0                         2.0   
207      1.0    1                 NaN    NaN                         NaN   
216      2.0    1                 NaN    NaN                         NaN   
219      1.0    1                 NaN   52.0                         3.0   
222      1.0    1                 NaN  112.0                         4.0   
227      1.0    2                 NaN    NaN                         NaN   
253      2.0    1                 NaN  112.0                         3.0   
265      2.0    1                 NaN   56.0                         3.0   
274      1.0    1                 NaN   76.0                         NaN   
282      1.0    1                 NaN   70.0                         3.0   
288      1.0    1                 NaN    NaN                         NaN   

     pain  outcome  
5     2.0      1.0  
16    4.0      2.0  
28    3.0      2.0  
35    4.0      3.0  
52    1.0      1.0  
56    3.0      1.0  
58    5.0      2.0  
74    3.0      2.0  
78    5.0      2.0  
87    2.0      3.0  
93    5.0      2.0  
96    2.0      1.0  
105   2.0      1.0  
110   2.0      2.0  
115   3.0      1.0  
136   5.0      2.0  
159   2.0      1.0  
160   4.0      1.0  
169   5.0      3.0  
174   4.0      2.0  
175   3.0      1.0  
177   4.0      2.0  
191   4.0      1.0  
207   3.0      2.0  
216   2.0      1.0  
219   3.0      1.0  
222   2.0      1.0  
227   3.0      1.0  
253   5.0      1.0  
265   3.0      3.0  
274   3.0      3.0  
282   2.0      2.0  
288   3.0      1.0

Заменим NaN в pulse на медиану, сгруппировав по возрасту и боли

In [147]:
horse_data.pulse.fillna(horse_data.groupby(['age', 'pain']).pulse.transform('median'), inplace = True)
horse_data.loc[horse_data.rectal_temperature.isnull()].head()

surgery  age  rectal_temperature  pulse  temperature_of_extremities  pain  \
5       2.0    1                 NaN   70.0                         2.0   2.0   
16      1.0    2                 NaN  128.0                         3.0   4.0   
28      1.0    1                 NaN   60.0                         NaN   3.0   
35      2.0    1                 NaN  104.0                         4.0   4.0   
52      2.0    1                 NaN   48.0                         1.0   1.0   

    outcome  
5       1.0  
16      2.0  
28      2.0  
35      3.0  
52      1.0

Предположим, что ректальная тем-ра связана с возрастом, наличием боли

In [149]:
horse_data.rectal_temperature.fillna(horse_data.groupby(['age', 'pain']).rectal_temperature.transform('median'), inplace = True)
horse_data.loc[horse_data.temperature_of_extremities.isnull()]

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
28       1.0    1              38.150   60.0                         NaN   
56       1.0    1              38.150   60.0                         NaN   
74       1.0    2              38.500  120.0                         NaN   
87       2.0    1              38.250   40.0                         NaN   
96       2.0    1              38.250   84.0                         NaN   
113      1.0    1              36.800   60.0                         NaN   
141      2.0    1              36.000   42.0                         NaN   
160      1.0    1              38.025   77.5                         NaN   
169      1.0    1              38.100   65.0                         NaN   
175      1.0    1              38.150   60.0                         NaN   
177      1.0    1              38.025   68.0                         NaN   
207      1.0    1              38.150   60.0                         NaN   
216      2.0    1              38.250   70.0                         NaN   
227      1.0    2              38.500  120.0                         NaN   
274      1.0    1              38.150   76.0                         NaN   
288      1.0    1              38.150   60.0                         NaN   

     pain  outcome  
28    3.0      2.0  
56    3.0      1.0  
74    3.0      2.0  
87    2.0      3.0  
96    2.0      1.0  
113   3.0      2.0  
141   2.0      2.0  
160   4.0      1.0  
169   5.0      3.0  
175   3.0      1.0  
177   4.0      2.0  
207   3.0      2.0  
216   2.0      1.0  
227   3.0      1.0  
274   3.0      3.0  
288   3.0      1.0

Оставшиеся пропуски в температуре конечностей заменим модой

In [150]:
horse_data.temperature_of_extremities.fillna(horse_data.temperature_of_extremities.round().mode()[0], inplace = True)
horse_data.loc[horse_data.temperature_of_extremities.isnull()]

Empty DataFrame
Columns: [surgery, age, rectal_temperature, pulse, temperature_of_extremities, pain, outcome]
Index: []

In [151]:
procent_isna(horse_data)

surgery - 0.0%
age - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
temperature_of_extremities - 0.0%
pain - 0.0%
outcome - 0.0%


In [152]:
  horse_data.describe() - horse_data_old

abdomcentesis_total_protein  age   outcome       pain      pulse  \
count                          NaN  0.0  1.000000  55.000000  24.000000   
mean                           NaN  0.0 -0.001839  -0.051020   0.393623   
std                            NaN  0.0 -0.000544  -0.104544  -0.684671   
min                            NaN  0.0  0.000000   0.000000   0.000000   
25%                            NaN  0.0  0.000000   0.000000   0.000000   
50%                            NaN  0.0  0.000000   0.000000   0.500000   
75%                            NaN  0.0  0.000000   0.000000   0.000000   
max                            NaN  0.0  0.000000   0.000000   0.000000   

       rectal_temperature   surgery  temperature_of_extremities  
count           60.000000  1.000000                   56.000000  
mean             0.027417 -0.001327                    0.026639  
std             -0.049091 -0.000282                   -0.038283  
min              0.000000  0.000000                    0.000000  
25%              0.000000  0.000000                    0.000000  
50%              0.000000  0.000000                    0.000000  
75%              0.000000  0.000000                    0.000000  
max              0.000000  0.000000                    0.000000

In [153]:
horse_data.mode() - horse_data_mode_old

abdomcentesis_total_protein  age  outcome  pain  pulse  rectal_temperature  \
0                          NaN  0.0      0.0   0.0   12.0                 0.0   
1                          NaN  NaN      NaN   NaN    NaN                 NaN   

   surgery  temperature_of_extremities  
0      0.0                         0.0  
1      NaN                         NaN

Сильных отклонений не выявлено